src/matrices.rs

```diff
@@ -2,217 +2,305 @@
-        }
+            !self
+                .data
+                .iter()
+                .skip(1)
+                .fold((false, self.data[0].len()), |(acc, row_len), next| {
+                    (acc || row_len != next.len(), row_len)
+                })
+                .0 // does any row have different length?
+        };
     }
 
-    fn checked_operation_on_two<F>(&self, other: &Self, operation: F) -> anyhow::Result<Self>
+    /// Returns the raw data of the matrix.
+    /// # Examples
+    /// ```rust
+    /// use matrix::Matrix;
+    /// let m = Matrix::new(vec![vec![1, 2], vec![3, 4]]).unwrap();
+    /// assert_eq!(m.get_data(), &vec![vec![1, 2], vec![3, 4]]);
+    /// ```
+    pub fn get_data(&self) -> &Vec<Vec<T>> {
+        &self.data
+    }
+
+    /// Returns the raw data of the matrix and consumes the matrix.
+    /// # Examples
+    /// ```rust
+    /// use matrix::Matrix;
+    /// let m = Matrix::new(vec![vec![1, 2], vec![3, 4]]).unwrap();
+    /// assert_eq!(m.into_data(), vec![vec![1, 2], vec![3, 4]]);
+    /// ```
+    pub fn consume(self) -> Vec<Vec<T>> {
+        self.data
+    }
+
+    /// Checks if two matrices have the same shape.
+    /// Shape of a matrix is a tuple of the form `(height, width)`.
+    /// # Arguments
+    /// * `other` - The other matrix.
+    /// # Returns
+    /// `true` if the matrices have the same shape, `false` otherwise.
+    /// # Examples
+    /// ```rust
+    /// use matrix::Matrix;
+    /// let m1 = Matrix::new(vec![vec![1, 2], vec![3, 4]]).unwrap();
+    /// let m2 = Matrix::new(vec![vec![1, 2], vec![3, 4]]).unwrap();
+    /// assert!(m1.has_same_shape(&m2));
+    /// let m1 = Matrix::new(vec![vec![1, 2], vec![3, 4]]).unwrap();
+    /// let m2 = Matrix::new(vec![vec![1, 2, 3], vec![4, 5, 6]]).unwrap();
+    /// assert!(!m1.has_same_shape(&m2));
+    /// ```
+    pub fn same_shapes(&self, other: &Self) -> bool {
+        let self_shape = self.get_shape();
+        let other_shape = other.get_shape();
+        self_shape == other_shape
+    }
+
+    /// Checks if two matrices can be multiplied.
+    /// Matrices can be multiplied if the number of columns of the first matrix
+    /// is equal to the number of rows of the second matrix.
+    /// # Arguments
+    /// * `other` - The other matrix.
+    /// # Returns
+    /// `true` if the matrices can be multiplied, `false` otherwise.
+    /// # Examples
+    /// ```rust
+    /// use matrix::Matrix;
+    /// let m1 = Matrix::new(vec![vec![1, 2], vec![3, 4], vec![5, 6]]).unwrap();
+    /// let m2 = Matrix::new(vec![vec![1, 2, 3], vec![4, 5, 6]]).unwrap();
+    /// assert!(m1.can_multiply(&m2));
+    /// let m1 = Matrix::new(vec![vec![1, 2], vec![3, 4]]).unwrap();
+    /// let m2 = Matrix::new(vec![vec![1, 2, 3], vec![4, 5, 6]]).unwrap();
+    /// assert!(!m1.can_multiply(&m2));
+    /// ```
+    pub fn corresponding_shapes_for_mul(&self, other: &Self) -> bool {
```

Comment:
Does this function need to be public? And the same question for `check_operation_on_two` and some other helper methods.

Reply:
You're right. This one doesn't sound useful for users, but as ::matrices can be used like a library it has some functions, that maybe useful.